src/main/java/com/sproutsocial/nsq/ListBasedBalanceStrategy.java

```diff
@@ -0,0 +1,131 @@
+package com.sproutsocial.nsq;
+
+import org.slf4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import static com.sproutsocial.nsq.Util.checkNotNull;
+import static org.slf4j.LoggerFactory.getLogger;
+
+public class ListBasedBalanceStrategy extends BasePubSub implements BalanceStrategy {
+    private static final Logger logger = getLogger(ListBasedBalanceStrategy.class);
+    protected final List<ConnectionDetails> daemonList;
+    private Publisher parent;
+    private Function<List<ConnectionDetails>, ConnectionDetails> connectionDetailsSelector;
+    private int failoverDurationSecs = 300;
+
+    /**
+     * Create a list based failover strategy that will alternate between all connected nsqd.  Will reconnect to a
+     * disconnected or failed nsqd on the next publish that could be routed to that nsqd after the failoverDuration has
+     * expired (Default 5 min).
+     * <p>
+     * This will throw an NSQD exception if all nsqd are in a failed state.
+     *
+     * @param nsqd a list of strings that represent HostAndPort objects.
+     */
+    public static BiFunction<Client, Publisher, BalanceStrategy> getRoundRobinStrategyBuilder(List<String> nsqd) {
+        return (c, p) -> buildRoundRobinStrategy(c, p, nsqd);
+    }
+
+    /**
+     * Create a list based failover strategy that shows strong preference to the first nsqd on the list.
+     * <p>
+     * On publish, find the first nsqd in this list that is in a connected or connectable state.  A nsqd is connectable
+     * if it has previously failed more than the configured failoverDuration (Default 5 min).
+     * <p>
+     * This will throw an NSQD exception if all nsqd are in a failed state.
+     *
+     * @param nsqd a list of strings that represent HostAndPort objects.
+     */
+    public static BiFunction<Client, Publisher, BalanceStrategy> getFailoverStrategyBuilder(List<String> nsqd) {
+        return (c, p) -> buildFailoverStrategy(c,p,nsqd);
+    }
+
+    private static ListBasedBalanceStrategy buildRoundRobinStrategy(Client client, Publisher parent, List<String> nsqd) {
+        return new ListBasedBalanceStrategy(client, parent, nsqd, new Function<List<ConnectionDetails>, ConnectionDetails>() {
+            private volatile int nextDaemonIndex = 0;
```

Comment:
Why the `volatile` here? Isn't the entire publish call guarded by a big `synchronized` block?

Reply:
Will remove.  This was a hold over when I was trying to reduce sync's...